eis_toolkit/raster_processing/unique_combinations.py

```diff
@@ -0,0 +1,62 @@
+from typing import List, Dict, Tuple
+
+import itertools
+import numpy as np
+import rasterio
+
+from eis_toolkit.exceptions import InvalidParameterValueException
+
+
+def _unique_combinations(
+    bands: List[np.ndarray],
+) -> np.ndarray:
+
+    out_image = np.empty_like(bands[0])
+    combinations: Dict[tuple, int] = {}
+    for row, column in itertools.product(range(len(bands[0])), range(len(bands[0][0]))):
+
+        combination = tuple(bands[band][row][column] for band, _ in enumerate(bands))
+
+        if combination not in combinations:
+            combinations[combination] = len(combinations) + 1
+        out_image[row][column] = combinations[combination]
+
+    return out_image
+
+
+def unique_combinations(  # type: ignore[no-any-unimported]
+    raster_list: List[rasterio.io.DatasetReader],
+) -> Tuple[np.ndarray, dict]:
+    """Get combinations of raster values between rasters.
+
+    All bands in all rasters are used for analysis.
+    A single band raster is used for reference when making the output.
+
+    Args:
+        raster_list (List[rasterio.io.DatasetReader]): Rasters to be used for finding combinations.
+
+    Returns:
+        out_image (numpy.ndarray): Combinations of rasters.
+        out_meta (dict): The metadata of the first raster in raster_list.
+    """
+    rasters = []
+    out_meta = raster_list[0].meta
+    out_meta['count'] = 1
+    height = raster_list[0].meta['height']
+    width = raster_list[0].meta['width']
+
+    for raster in raster_list:
+        for band in range(1, raster.count+1):
+            rasters.append(raster.read(band))
```

Comment:
It is quite easy to run out of memory when reading rasters fully into memory as `numpy` arrays. But as long as the rasters are small, this is not a concern. Just something to keep in mind.

Reply:
Thanks for the info. Is there another way to read the rasters and use them in the function?

Reply:
@jtlait uses windowed reads here already in `eis_toolkit`: https://github.com/GispoCoding/eis_toolkit/blob/77fba1fc64755f898d7849bb79c7b86df6b3211d/eis_toolkit/raster_processing/windowing.py#L11

Most relevant lines:

~~~python
    window = Window(
        col_off=top_left_col,
        row_off=top_left_row,
        width=width,
        height=height,
    )

    out_image = raster.read(
        boundless=True,
        window=window,
        fill_value=-9999,
    )
~~~

The basic principle is to read only a "window" of the raster into an array. I could not readily say how easy it is to implement for the functionality you are doing here. It is more difficult to work with windows of data and then in the end make sure you have correctly aggregated the results from all the windows.

Reply:
Thank you for the tip. I'll keep this in mind as I am not sure how to implement this, yet.

Reply:
Returning to this PR and conversation, I think we should, for now, go with the initial implementation of reading the raster fully into memory. There are other functions already present in the EIS Toolkit that read rasters similarly and don't take these extra measures for memory management, so if we decide to take them (maybe find out in the testing that it is necessary), then we can update multiple problematic functions at once.

Reply:
I have now made some optimizations to the code. With the `data/remote/workflow_demo/IOCG_AEM_Inph_.tif` file I got the execution from ~4 seconds to ~2 seconds. However, there is a triple nested loop in the new implementation that could cause slowness with bigger datasets.